mne/beamformer/_ap.py

```diff
@@ -0,0 +1,712 @@
+    return s_ap_2
+
+
+def _calculate_fixed_alternating_projections(data_arr, gain,
+                                             nsources,
+                                             max_iter):
+    """Calculate fixed-orientation alternating projection.
+
+    Parameters
+    ----------
+    data_arr : array, shape (nchannels, times)
+        Filtered evoked data array.
+    gain : array, shape (nchannels, ndipoles)
+        Gain matrix.
+    nsources : int
+        The number of dipoles to estimate.
+    max_iter : int
+        Maximal iteration number of AP.
+
+    Returns
+    -------
+    s_ap_2 : list of int
+        List of dipole indices.
+
+    """
+    s_ap = []
+    ndipoles = gain.shape[1]
+    attr_dict = {
+        'ndipoles': ndipoles,
+        'nsources': nsources,
+        'max_iter': max_iter
+    }
+    logger.info('calculating fixed-orientation alternating projection')
+    data_cov = _produce_data_cov(data_arr, attr_dict)
+
+    # ######################################
+    # 1st Phase
+    # (a) Initialization: search the 1st source location over the entire
+    # dipoles topographies space
+    # ######################################
+
+    logger.info(' 1st phase : ')
+    s_ap = _fixed_phase1a(attr_dict, data_cov, gain)
+
+    # ######################################
+    # (b) Now, add one source at a time
+    # ######################################
+
+    s_ap = _fixed_phase1b(gain, s_ap, data_cov, attr_dict)
+
+    # #####################################
+    # 2nd phase
+    # #####################################
+
+    logger.info(' 2nd phase : ')
+    s_ap_2 = copy(s_ap)
+    s_ap_2 = _fixed_phase2(attr_dict, s_ap_2, gain, data_cov)
+
+    return s_ap_2
+
+
+def _solve_active_gain_eig(ind, gain, data_cov, eig, perpend_spc):
+    """Eigen values and vector of the projection."""
+    gain_idx = list(range(ind * 3, ind * 3 + 3))
+    l_p = gain[:, gain_idx]
+    eig_a = multi_dot(
+        [l_p.T, perpend_spc, data_cov, perpend_spc, l_p])
+    eig_b = multi_dot(
+        [l_p.T, perpend_spc, perpend_spc, l_p])
+    eig_b = eig_b + 1e-3 * eig_b.trace() * np.eye(3)
+    eig_val, eig_vec = eig(eig_a, eig_b)
+
+    return eig_val, eig_vec, l_p
+
+
+def _free_phase1a(attr_dict, gain, data_cov):
+    """Calculate phase 1a of free oriented AP.
+
+    Initialization: search the 1st source location over
+    the entire dipoles topographies space.
+
+    Parameters
+    ----------
+    attr_dict : dict
+        See: _calculate_free_alternating_projections.
+    data_cov : array
+        Data Covariance.
+    gain : array, shape (nchannels, 3 * ndipoles)
+        Gain matrix.
+
+    Returns
+    -------
+    s_ap : list of int
+        List of dipole indices.
+    oris : array, shape (nsources, 3)
+        Orientations array of estimated sources (sorted by s_ap).
+    sub_g_proj : array
+        Sub space projected by estimated dipoles.
+
+    """
+    from scipy.linalg import eig
```

Comment:
it should be `from scipy import linalg` at the top of the file and then inside `linalg.eig`

see how it's done in other parts of mne.